src/Dscp/DB/SQLite/Instances.hs

```diff
@@ -1,28 +1,63 @@
-import Dscp.Core.Serialise ()
-import Dscp.Educator.Txs (PrivateTxId)
-
--- TODO(kir): split into separate .Instances module.
-instance FromField Address where
-    fromField f = Codec.deserialise <$> fromField f
-
-instance FromField CourseId where
-    fromField f = Codec.deserialise <$> fromField f
 
-instance FromField Grade where
-    fromField f = Codec.deserialise <$> fromField f
-
-instance FromField SubmissionSig where
-    fromField f = Codec.deserialise <$> fromField f
-
-instance ToField PrivateTxId where
-    toField = toField . serialise
+import Dscp.Core.Serialise ()
+import Dscp.Core.Types (Address (..), Assignment (..), AssignmentType, CourseId (..), Grade (..),
+                        SignedSubmission (..), SubjectId (..), Submission (..), SubmissionType,
+                        SubmissionWitness (..))
+import Dscp.Crypto (Hash, PublicKey, Signature, hash)
+import Dscp.Educator.Txs (PrivateTx (..))
+
+instance FromField (Hash a)       where fromField f = Codec.deserialise <$> fromField f
+instance FromField (Signature a)  where fromField f = Codec.deserialise <$> fromField f
+
+-- TODO(kir): use #define to generate macros
+instance FromField Address        where fromField f = Codec.deserialise <$> fromField f
+instance FromField PublicKey      where fromField f = Codec.deserialise <$> fromField f
+instance FromField SubjectId      where fromField f = SubjectId         <$> fromField f
+instance FromField CourseId       where fromField f = CourseId          <$> fromField f
+instance FromField Grade          where fromField f = Codec.deserialise <$> fromField f
+instance FromField AssignmentType where fromField f = Codec.deserialise <$> fromField f
+instance FromField SubmissionType where fromField f = Codec.deserialise <$> fromField f
+
+instance ToField   (Hash a)       where toField     = toField . Codec.serialise
+instance ToField   (Signature a)  where toField     = toField . Codec.serialise
+
+instance ToField   Address        where toField     = toField . Codec.serialise
+instance ToField   CourseId       where toField     = toField . getCourseId
+instance ToField   AssignmentType where toField     = toField . Codec.serialise
+instance ToField   SubmissionType where toField     = toField . Codec.serialise
+
+instance FromRow   CourseId       where fromRow     = field
+instance FromRow   Grade          where fromRow     = field
+
+instance FromRow Assignment where
+    fromRow = Assignment <$> field <*> field <*> field
+
+instance ToRow Assignment where
+    toRow task@ (Assignment course ty text) =
+        [toField (hash task), toField course, toField ty, toField text]
+
+instance FromRow Submission where
+    fromRow = Submission <$> field <*> field <*> fromRow
+
+-- These three structures depend on the public key, which is not in db.
+-- Therefore, we have to plug it in somehow.
+instance FromRow (PublicKey -> SubmissionWitness) where
+    fromRow = (\sig pk -> SubmissionWitness pk sig) <$> field
```

Comment:
Ueee, finially I see instance for a function written in production code :man_dancing:

Reply:
Fix'd.